lfs.c

```diff
@@ -3189,6 +3189,82 @@ static int lfs_rawstat(lfs_t *lfs, const char *path, struct lfs_info *info) {
     return lfs_dir_getinfo(lfs, &cwd, lfs_tag_id(tag), info);
 }
 
+#ifndef LFS_READONLY
+static int lfs_dir_prep_removal(lfs_t *lfs, struct lfs_mlist *dir,
+        lfs_mdir_t *newcwd, uint16_t newid, lfs_block_t *pair,
+        lfs_gstate_t *tmp_gstate)
+{
+    lfs_stag_t res = lfs_dir_get(lfs, newcwd, LFS_MKTAG(0x700, 0x3ff, 0),
+            LFS_MKTAG(LFS_TYPE_STRUCT, newid, 8), pair);
+    if (res < 0) {
+        return (int)res;
+    }
+    lfs_pair_fromle32(pair);
+
+    int err = lfs_dir_fetch(lfs, &dir->m, pair);
+    if (err) {
+        return err;
+    }
+
+    if (dir->m.count > 0 || dir->m.split) {
+        uint16_t id = 0;
+
+        // Walk tags stored in this directory and check for any directory
+        // tags.  Removal of directories with a directory in them can lead
+        // to additional orphans in the filesystem, so we return
+        // LFS_ERR_NOTEMPTY in this case.  Otherwise, leave the loaded
+        // directory for the tail end of the directory split to leave a proper
+        // view of the filesystem after removal.
+        while (true) {
+            if (dir->m.count == id) {
+                if (!dir->m.split) {
+                    // We have iterated through the folder to the last
+                    // tag.
+                    break;
+                }
+
+                // Before we fetch the next block, update our fetched gstate xor
+                lfs_dir_getgstate(lfs, &dir->m, tmp_gstate);
+
+                err = lfs_dir_fetch(lfs, &dir->m, dir->m.tail);
+                if (err) {
+                    return err;
+                }
+
+                id = 0;
+            }
+
+            lfs_stag_t tag = lfs_dir_get(lfs, &dir->m, LFS_MKTAG(0x780, 0x3ff, 0),
+                    LFS_MKTAG(LFS_TYPE_NAME, id, 0), NULL);
+
+            if (tag < 0) {
+                return tag;
+            }
+
+            if (lfs_tag_type3(tag) == LFS_TYPE_DIR) {
+                return LFS_ERR_NOTEMPTY;
```

Comment:
Assuming this is moved to `lfs_removeall` or equivalent: Perhaps this should return ENOSYS for now (feel free to add `LFS_ERR_NOSYS = -38` to the lfs_error enum). This suggests the error code may change.

It would be nice to add support for recursive removal of directories in the future if possible.

Reply:
If I'm understanding the filesystem correctly, the linked list of "directories" is only contiguous for one directory at a time?  E.g. a "split" of a given directory will have all of its links together?  Recursive support would potentially require multiple touch points to the linked list structure, but a single folder without subfolders doesn't have this requirement.

However, one could remove entire directories together at a "super" function.  That function would need to be semi-recursive.  It could keep looking for folders, removing those, and then starting over at the top again.  (That way it doesn't have as much of an in-memory recursive footprint and depend on the on-disk state.)

I did also add support in the removal by move, too.  New function name for this, too?

Reply:
> If I'm understanding the filesystem correctly, the linked list of "directories" is only contiguous for one directory at a time? E.g. a "split" of a given directory will have all of its links together?

Yep, directories can't be interleaved.

> Recursive support would potentially require multiple touch points to the linked list structure, but a single folder without subfolders doesn't have this requirement.

`lfs_fs_deorphan` already handles multiple orphans spread out in the linked-list (can be created by wear-leveling/bad-block removal), so I don't think that will be a problem. The main challenge is how to determine if a metadata-pair is an orphan, currently `lfs_fs_deorphan` only looks up the immediate parent, and is already expensive.

> It could keep looking for folders, removing those, and then starting over at the top again. (That way it doesn't have as much of an in-memory recursive footprint and depend on the on-disk state.)

This might work, if you also remove the directory entry to keep track of progress. The goal of littlefs is fixed RAM cost, so really no recursion is allowed at all (I'm being a bit hypocritical at the moment, but working to remove the one recursive function in the codebase). But and tail-recursive function can be written as a loop, so it sounds like it should work.

> I did also add support in the removal by move, too. New function name for this, too?

Oh I missed that. Hmm, that's a good question, `lfs_renameall` certainly doesn't make sense.

Maybe `lfs_removerecursive` and `lfs_renamerecursive`?

As an aside that is certainly quite a bit of functionality for one little function.

Feel free to refactor the internals to reuse code between the recursive/POSIX versions.

Reply:
> Maybe `lfs_removerecursive` and `lfs_renamerecursive`?

Hah, yeah.  Before seeing this, I called it "lfs_rename_with_removeall" which is _really_ verbose (and probably too verbose), but does describe it accurately.  I'm not sure I care for renamerecursive personally, but I'm not sure I really care for what I called it either.
 
> As an aside that is certainly quite a bit of functionality for one little function.
> 
> Feel free to refactor the internals to reuse code between the recursive/POSIX versions.

Yeah, I modified the rawremove(...) to accept a pointer to the bit of code that does the work in the "recursive" side of things.  I figured this reuses code between the two fairly easily; is this an okay approach for you?  The code will be dropped by a linker with garbage collection if noone references the helper function to put it into an argument.

Thanks,
Tim